content/en/docs/reference/access-authn-authz/webhook.md

```diff
@@ -121,6 +121,24 @@ To disallow access, the remote service would return:
 }
 ```
 
+A special case is when an authorization webhook can't make a decision based on
+the information available. By setting both status attributes `allowed` and `denied`
+to `false` the webhook signals "no opinion",  which will forward the request to the
```

Comment:
This isn't quite accurate. Here are the three possible status responses:

* Allowed: `{"allowed": true}`
* Forbidden: `{"allowed": false, "denied": true}`
* No opinion: `{"allowed": false}`

If either "allowed" or "denied" are absent, they are treated as `false`.

Reply:
> A special case is when an authorization webhook can't make a decision based on the information available.

Actually, "no opinion" is the common case. Denied is the special case, since it requires the authorizer to be aware of how it is configured together with other authorizers.

Reply:
Thanks for responding so quickly @liggitt. What you are saying however seems unsupported both by the docs in the code on [SubjectAccessReview](https://github.com/kubernetes/kubernetes/blob/74618928c25302943a3a5e903bf9a3626bc1e84a/pkg/apis/authorization/types.go#L134) (see especially "Denied") and the [switch statement](https://github.com/kubernetes/kubernetes/blob/77fca12f66c7ca0a6f62da1a18fd61ac6f4cc1d5/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L208-L217) making the Authorize call. Granted I'm new in this context so might be I'm misinterpreting.

On the "common case" - the docs seem to be quite clear [elsewhere](https://kubernetes.io/docs/reference/access-authn-authz/authorization/#determine-whether-a-request-is-allowed-or-denied) that the no opinion case is what will have the next authorizer consulted - denying the request does not forward it to the next authorizer.

> When multiple authorization modules are configured, each is checked in sequence. If any authorizer approves or denies a request, that decision is immediately returned and no other authorizer is consulted. If all modules have no opinion on the request, then the request is denied.

Reply:
There are four possible combinations:

| | allowed=true | allowed=false |
|-|---|---|
| denied=true | [invalid](https://github.com/kubernetes/kubernetes/blob/77fca12f66c7ca0a6f62da1a18fd61ac6f4cc1d5/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L209-L210) | [denied](https://github.com/kubernetes/kubernetes/blob/77fca12f66c7ca0a6f62da1a18fd61ac6f4cc1d5/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L211-L212) |
| denied=false | [allowed](https://github.com/kubernetes/kubernetes/blob/77fca12f66c7ca0a6f62da1a18fd61ac6f4cc1d5/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L213-L214) | [no opinion](https://github.com/kubernetes/kubernetes/blob/77fca12f66c7ca0a6f62da1a18fd61ac6f4cc1d5/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L215-L216) |

Reply:
> On the "common case" - the docs seem to be quite clear elsewhere that the no opinion case is what will have the next authorizer consulted - denying the request does not forward it to the next authorizer.

The currently documented response (`{"allowed": false}`) produces a `no opinion` result. That is the normal response authz webhooks should return when they do not want to authorize a request.

The "deny and short-circuit" response (`{"denied": true}`) is a special case that should only be returned when a webhook has knowledge about how it is configured together with other authorizers, and knows that no later authorizers should be consulted.

Reply:
While leaving out `{"denied": false}` will have the same effect as stating it (as it will default to `false` if not provided) - the ambiguity of stating both `{"allowed": false}` _and_ `{"denied": false}` makes it a lot more fit for documenting the "no opinion" case IMO. The docs currently document this case as:

> To disallow access, the remote service would return [ .. the implied "no opinion" response ]

As a webhook author you'd be surprised to see users authorized after having "disallowed access" for them, no?

Given the "implied no opinion" case this doc PR will need more work - if you think it's worthwhile I'll be happy to spend some time on this tomorrow taking your feedback above into consideration.

Reply:
> As a webhook author you'd be surprised to see users authorized after having "disallowed access" for them, no?

As a webhook author, perhaps, but not as a cluster administrator who had configured multiple authorizers, one of which allowed that request.

I agree clarification of the normal case would be beneficial, and documentation of the denied case is needed, so efforts to improve these docs are appreciated.

I want to ensure that webhook authors understand the response of a webhook that does not want to authorize a response should typically be the "no opinion" response. That allows the webhook to interoperate with other configured authorizers well.

In more exceptional cases where a webhook has more complete knowledge of the cluster it is deployed into and all other authorizers, returning a denied response could be appropriate, but the docs should make it clear that is not typical.

Reply:
Thanks a lot @liggitt for taking the time - really appreciate it. I'll set aside tomorrow to update these docs trying to make all the three cases clear, including when one might want to use each and why "no opinion" should be the preferred case if uncertain. Not sure if it's possible to add new edits on top of one of these PR's started right on Github but I'll try. If not I'll open a new one.